File: lab7/1.cpp
#include <cmath>
#include <iostream>
#include <vector>

// Функция для вычисления производной dy/dx
double dydx(double y, double x) { return -y + std::exp(x); }

// Функция для вычисления точного решения
double exactSolution(double x) { return 0.5 * std::exp(x) + std::exp(-x); }

int main() {
  double h = 0.1;  // Шаг интегрирования
  double x0 = 0.0; // Начальное значение x
  double y0 = 1.5; // Начальное значение y
  double xn = 1.0; // Конечное значение x

  // Векторы для хранения значений x, y, точного решения и ошибки
  std::vector<double> x;
  std::vector<double> y;
  std::vector<double> y_exact;
  std::vector<double> error;

  // Инициализация начальных значений
  x.push_back(x0);
  y.push_back(y0);
  y_exact.push_back(exactSolution(x0));
  error.push_back(std::abs(y0 - y_exact.back()));

  int n = static_cast<int>((xn - x0) / h); // Количество шагов

  // Основной цикл метода Эйлера
  for (int i = 0; i < n; ++i) {
    double x_next = x[i] + h;
    double y_next = y[i] + h * dydx(y[i], x[i]);

    x.push_back(x_next);
    y.push_back(y_next);
    y_exact.push_back(exactSolution(x_next));
    error.push_back(std::abs(y_next - y_exact.back()));
  }

  // Вывод результатов
  std::cout << "x\ty (Эйлер)\ty (Точное)\tРазница\n";
  for (int i = 0; i < x.size(); ++i) {
    std::cout << x[i] << "\t" << y[i] << "\t" << y_exact[i] << "\t" << error[i]
              << "\n";
  }

  return 0;
}

File: lab7/1.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


# Функция для вычисления правой части дифференциального уравнения
def f(x, y):
    return -y + np.exp(x)


# Начальные условия
x0 = 0
y0 = 1.5
h = 0.1
xn = 1

# Число шагов
n = int((xn - x0) / h)

# Массивы для хранения значений x и y
x = np.zeros(n + 1)
y = np.zeros(n + 1)
x[0] = x0
y[0] = y0

# Реализация метода Эйлера
for i in range(n):
    y[i + 1] = y[i] + h * f(x[i], y[i])
    x[i + 1] = x[i] + h

# Точное решение
y_exact = (np.exp(x) / 2) + np.exp(-x)

# Расчёт погрешности
errors = np.abs(y - y_exact)

# Создание таблицы
data = {"x": x, "y (Euler)": y, "y (Exact)": y_exact, "Error": errors}
df = pd.DataFrame(data)

# Вывод таблицы
print(df)

# Построение графиков
plt.plot(x, y, "o-", label="Метод Эйлера")
plt.plot(x, y_exact, "-", label="Точное решение")
plt.xlabel("x")
plt.ylabel("y")
plt.legend()
plt.title("Решение задачи Коши методом Эйлера")
plt.show()
